RESULT_PROPOSAL.md

```diff
@@ -0,0 +1,295 @@
+                return;
+            }
+            // unexpected error - one of nonce or error should not be null
+        }));
+
+payPalClient.createPaymentAuthRequest(activity, request, (paymentAuthRequest, error) -> {
+    if (error != null) {
+        // handle error
+        return;
+    }
+    if (paymentAuthRequest != null) {
+        payPalLauncher.launch(activity, paymentAuthRequest);
+        return;
+    }
+    // unexpected error - one of paymentAuthRequest or error should not be null
+});
+```
+
+## Listener
+
+This integration aligns most closely with the v4 integration listener pattern, but either
+requires two separate listeners (for launcher and client), or couples the launcher and the
+client because both need to be invoked in different listener methods. The benefit of a single
+listener is that errors throughout the entire flow can be handled in one method (rather than two
+separate error handling points with the callback patterns).
+
+### Kotlin
+
+```kotlin
+payPalLauncher = PayPalLauncher(listener) 
+payPalClient = PayPalClient(listener)
+payPalClient.createPaymentAuthRequest(activity, request)
+
+override fun onPaymentAuthRequest(paymentAuthRequest: PaymentAuthRequest) {
+    payPalLauncher.launch(paymentAuthRequest)
+}
+
+override fun onPaymentAuthResult(paymentAuthResult: PaymentAuthResult) {
+    payPalClient.tokenize(paymentAuthResult)
+}
+
+override fun onSuccess(payPalAccountNonce: PayPalAccountNonce) {
+    // send nonce to server
+}
+
+override fun onCancel() {
+    // handle cancel
+}
+
+override fun onError(error: Exception) {
+    // handle error
+}
+```
+
+### Java
+
+```java
+PayPalLauncher payPalLauncher = new PayPalLauncher(listener);
+PayPalClient payPalClient = new PayPalClient(listener);
+payPalClient.createPaymentAuthRequest(activity, request);
+
+@Override
+public void onPaymentAuthRequest(PaymentAuthRequest paymentAuthRequest) {
+    payPalLauncher.launch(paymentAuthRequest)
+}
+
+@Override
+public void onPaymentAuthResult(PaymentAuthResult paymentAuthResult) {
+    payPalClient.tokenize(paymentAuthResult)
+}
+
+@Override
+public void onSuccess(PayPalAccountNonce payPalAccountNonce) {
+    // send nonce to server
+}
+
+@Override
+public void onCancel() {
+    // handle cancel
+}
+
+@Ovrride
+public void onError(Exception error) {
+    // handle error
+}
+```
+## Single Result Object With Types
+
+This approach is the most Kotlin-first pattern. It also resolves the nullability issues. However, 
+the Java integration becomes somewhat complex with casting. This approach aligns with how other 
+payment SDKs handle results.
+
+This approach relies on the Kotlin sealed class, so requires the `PaymentAuthResult` and
+`PaymentResult` objects to live in `BraintreeCore`. It also requires the launchable result 
+objects to all extend a single class (ex: `LaunchableRequest`). This reduces code duplication in 
+our SDK, but requires casting by the merchants (ex: PaymentMethodNonce is returned instead of
```

Comment:
> requires casting by the merchants

Will this casting still be required once fully converted to Kotlin?

Reply:
Yes but Kotlin has [smart casting](https://kotlinlang.org/docs/typecasts.html#smart-casts) so for Kotlin merchants it's as simple as checking if the object `is PaymentResult.Success` and then it's automatically cast to that type. So the Kotlin code snippet below won't change but it's already relatively clean. 

Java merchants will always have the cumbersome type-checking/unwrapping.